### ISSUE TITLE
SMTP Error Self-Signed Certificate

### ISSUE THREAD

**Issue by camps7ee** [fmartinou/whats-up-docker#92]
I'm running Protonmail-Bridge as my SMTP server (it's a complicated little piece protonmail developed) and it runs of a self-signed TLS certificate for 127.0.0.1.

I'm trying to find where additional SMTP settings for WUD can be configured, cause it would be great to add a `WUD_SMTP_SERVER_TLS_SKIP_VERIFY` environment variable, or something similar to that, so that it can push through the connection.

For example, when you configure the SMTP Server in Thunderbird it allows you to confirm the security exception for that self-signed certificate (see https://protonmail.com/bridge/thunderbird#11).

Maybe this is already possible somehow?

Thank you in advance!

**Comment by fmartinou** [comment]
Hi,

Apparently, the underlying lib used for the SMTP trigger is supporting options to ignore the validation of the smtp server certificate:
https://nodemailer.com/smtp/#tls-options

**Comment by camps7ee** [comment]
Ah yeah, fair enough!

I was actually trying to find the code for the container to see if maybe I can help lay some foundational work for the change but couldn't find the `node.js` file.

In an attempt to help I want to mention to pay attention to the AUTH methods _nodemailer_ supports.

I'm mentioning it because _watchtower_ (as an example) uses [go.dev](https://pkg.go.dev/net/smtp?utm_source=godoc#PlainAuth) for their mail-delivery and it only works with `AUTH PLAIN` and `CRAMMD5`. However, to run `AUTH PLAIN` you cannot skip the verification of TLS, and if you do, the SMTP server needs to be running as `localhost`.

I was looking at the documentation you shared and it seems that with _nodemailer_ the default auth option is `AUTH LOGIN` and the second option is `oAUTH2` - It does not mention anything about localhosts or not though, so it seems it should be possible(?).

Excuse the rant and thank you again for a very cool service you've developed! :)

**Comment by fmartinou** [comment]
Thanks for your help and your support 😃 .

The code for the Smtp trigger is located here ([/app/triggers/providers/smtp/Smtp.js](https://github.com/fmartinou/whats-up-docker/blob/master/app/triggers/providers/smtp/Smtp.js)).

If you want to test locally wud, you need to:
1. Clone the repository
2. Install the dependencies `cd app && npm ci`
3. Run the app `cd app && npm start` (you can configure wud using [/app/nodemon.json](https://github.com/fmartinou/whats-up-docker/blob/master/app/nodemon.json))

_(In [/test/docker-compose.yml](https://github.com/fmartinou/whats-up-docker/blob/master/test/docker-compose.yml) you also have a docker-compose file to easily check)_

**Comment by camps7ee** [comment]
So I'm not really experienced with any coding tbh, nor how to properly be able to test this locally. I did however consult a friend of mine on possible changes to the `smpt.js` file that might help us achieve the objective. For whenever you find the time, here's the sample code he suggested:

```
 initTrigger() {
        this.transporter = nodemailer.createTransport({
            host: this.configuration.host,
            port: this.configuration.port,
            secure: this.configuration.secure,
            auth: {
                user: this.configuration.user,
                pass: this.configuration.pass,
            },
            tls: {
                // do not fail on invalid certs
                rejectUnauthorized: this.configuration.tls.rejectUnauthorized,
              },
        });
    }
```

If you're up for a chat whenever, I'm on discord `camps7ee#6056`

**Comment by fmartinou** [comment]
Hi,

I've already started working on it 😁

[Here is the main change I did](https://github.com/fmartinou/whats-up-docker/blob/feature/%2392/app/triggers/providers/smtp/Smtp.js#L51) (pretty similar to what you're proposing).

Could you try it by running `fmartinou/whats-up-docker:feature__92` and setting the variable `WUD_TRIGGER_SMTP_{trigger_name}_TLS_VERIFY=false` ?

**Comment by camps7ee** [comment]
Amazing :)

I just gave it a go and still error:

`13:47:22.647  WARN whats-up-docker/trigger.smtp.email: Error (self signed certificate)`

I'm suspecting that you need to include the `secure` variable as well so that it can be set to `true` and not fail on invalid certificates: https://nodemailer.com/smtp/#3-allow-self-signed-certificates

Either that or maybe `ignoreTLS` variable based of this part of the documentation:
`ignoreTLS – if this is true and secure is false then TLS is not used even if the server supports STARTTLS extension`

**Comment by fmartinou** [comment]
Let's try with the `secure` option... 😁 
https://github.com/fmartinou/whats-up-docker/blob/feature/%2392/app/triggers/providers/smtp/Smtp.js#L52

The fix is [currently building on travis](https://app.travis-ci.com/github/fmartinou/whats-up-docker/builds/243692207)

Once terminated, you just need to:
1. `docker pull fmartinou/whats-up-docker:feature__92` to overwrite the old image
2. set `WUD_TRIGGER_SMTP_{trigger_name}_TLS_ENABLED=true`
3. set `WUD_TRIGGER_SMTP_{trigger_name}_TLS_VERIFY=false`
4. recreate the container 🤞

**Comment by camps7ee** [comment]
okay okay, we're moving forward it seems, new error:

`Some triggers failed to register (Error when registering component smtp ("tls.enabled" is not allowed))`

**Comment by fmartinou** [comment]
The code seems ok 🤔 :
https://github.com/fmartinou/whats-up-docker/blob/feature/%2392/app/triggers/providers/smtp/Smtp.js#L21

`tls.enabled` should be seen as a valid option

Are you sure you run the latest image and not the previous one?
(because I overwritten the `docker:feature__92` tag)

Did you force the pull?

**Comment by camps7ee** [comment]
yeah, forced the pull just now and it's working 👍 :)

Now I got a different error though, not at all related to WUD, it's related to god damn protonmail-bridge and it's SSL:

`Error (140709971589936:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332`

Can WUD run on docker-compose version `2.1` or does it have to run `version 3` ?

**Comment by fmartinou** [comment]
It should be working on docker-compose `2.1` without any problem.

**Comment by camps7ee** [comment]
alright, yeah, something funky going on with protonmail-bridge....the configurations you've done are working great, amazing work :)

**Comment by fmartinou** [comment]
Great! 👍 
I close the issue and merge the fix in the `master` branch (it will be available in the next release).
If you encounter new problems, feel free to reopen the issue or open a new one.

**Comment by fmartinou** [comment]
Available in `5.12.0`.
